### PR TITLE
3 blog crud with role based control

### PR DESCRIPTION
**Pull request by Ismoil-JS**
create blog ==>
Screenshot 2024-10-08 at 14 15 56
get blog ==>
Screenshot 2024-10-08 at 14 23 01
get blog by id ==>
Screenshot 2024-10-08 at 14 23 07
patch blog ==>
Screenshot 2024-10-08 at 14 23 15
delete blog ==>
Screenshot 2024-10-08 at 14 23 21

### CODE REVIEW

File: src/entities/Blog.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  PrimaryGeneratedColumn,
+} from 'typeorm'
+
+@Entity()
+export class Blog {
+  @PrimaryGeneratedColumn('uuid')
+  id!: string
+
+  @Column()
+  author_id!: string
```
Comment (mukhammadjon-ss) [inline src/entities/Blog.ts:16]: Here you have to connect author and blog with foreign key. It is good for security and retrieving data from database. User and its articles and article with its user. Hint: use @OneToMany, @ManyToOne decorators

File: src/entities/Blog.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+  @Column()
+  title!: string
+
+  @Column()
+  content!: string
+
+  @Column('text', { array: true })
+  tags!: string[]
```
Comment (mukhammadjon-ss) [inline src/entities/Blog.ts]: tags are supposed to be optional

File: src/modules/blog/blog.service.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import { BlogUpdateDto } from './dtos/update-blog.dto'
+import { BlogParamsDto } from './schemas'
+
+export class BlogService {
+  private blogRepository = AppDataSource.getRepository(Blog)
+
+  async createBlog(payload: BlogCreateDto): Promise<Blog> {
+    const newBlog = new Blog()
+
+    newBlog.author_id = payload.author_id
```
Comment (mukhammadjon-ss) [inline src/modules/blog/blog.service.ts]: in previous task. You implemented authentication and authorization. You can get user from request. by its accessToken. it shouldn't come from request body. extract user from token and assign it here.

File: src/modules/blog/blog.service.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      skip: (page - 1) * limit,
+    })
+  }
+
+  async getBlogById(id: string): Promise<Blog | null> {
+    const blog = await this.blogRepository.findOne({
+      where: { id },
+    })
+
+    if (!blog) {
```
Comment (mukhammadjon-ss) [inline src/modules/blog/blog.service.ts:43]: here you have to throw proper not found error

File: src/modules/blog/dtos/create-blog.dto.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import { IsNotEmpty, IsString, IsUUID } from 'class-validator'
+
+export class BlogCreateDto {
+  @IsString()
+  @IsNotEmpty()
+  @IsUUID()
+  author_id!: string
```
Comment (mukhammadjon-ss) [inline src/modules/blog/dtos/create-blog.dto.ts]: here you don't need author id as you can get it from user token

File: src/modules/blog/dtos/create-blog.dto.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  @IsString()
+  @IsNotEmpty()
+  title!: string
+
+  @IsString()
+  @IsNotEmpty()
+  content!: string
+
+  @IsString({ each: true })
+  tags!: string[]
```
Comment (mukhammadjon-ss) [inline src/modules/blog/dtos/create-blog.dto.ts]: this should be optional

File: src/modules/blog/blog.controller.ts
Comment (mukhammadjon-ss) [inline src/modules/blog/blog.controller.ts:1]: when using async functions don't forget about error handling. first throw error from service and catch it in controller using trycatch block

**Review by mukhammadjon-ss — CHANGES_REQUESTED** [review]

File: src/entities/Blog.ts (hunk 1 above)
Comment (Ismoil-JS) [reply src/entities/Blog.ts:16]: when i create User entity it did not have blogs field. To correct my mistake on Blog entity I added blogs filed on User entity.

File: src/entities/Blog.ts (hunk 2 above)
Comment (Ismoil-JS) [reply src/entities/Blog.ts]: I did it optinal by using question mark

File: src/modules/blog/blog.controller.ts
Comment (Ismoil-JS) [reply src/modules/blog/blog.controller.ts:1]: implement trycatch

File: src/modules/blog/dtos/create-blog.dto.ts (hunk 6 above)
Comment (Ismoil-JS) [reply src/modules/blog/dtos/create-blog.dto.ts]: did it with question mark

File: src/modules/blog/blog.service.ts (hunk 4 above)
Comment (Ismoil-JS) [reply src/modules/blog/blog.service.ts:43]: I created a NotFoundError method

**Pull request merged**